core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -368,35 +407,44 @@ public void onFailure(Throwable e) {
 
         @Override
         protected void doRun() {
-            if (checkBlocks() == false) {
+            // shardID added to the request, as a side effect of calling resolveRequest
+            if (checkBlocksAndResolveRequest() == false) {
```

Comment:
I agree it would be clearer to perform the block checks and resolve request in the `doRun` method itself, https://github.com/areek/elasticsearch/commit/1159d3ead370dcc94835b5b5146f7344eef65a0a fixes this and makes it clear that we only operate on one copy of the state throughout a single run of the reroute phase.